ncclient/transport/tls.py

```diff
@@ -0,0 +1,145 @@
+
+DEFAULT_TLS_NETCONF_PORT = 6513
+DEFAULT_TLS_TIMEOUT = 120
+
+BUF_SIZE = 4096
+
+
+class TLSSession(Session):
+
+    def __init__(self, device_handler):
+        capabilities = Capabilities(device_handler.get_capabilities())
+        Session.__init__(self, capabilities)
+        self._host = None
+        self._connected = False
+        self._socket = None
+        self._buffer = StringIO()
+        self._device_handler = device_handler
+        self._message_list = []
+        self._closing = threading.Event()
+        self.parser = DefaultXMLParser(self)
+        self.logger = SessionLoggerAdapter(logger, {'session': self})
+
+    def _dispatch_message(self, raw):
+        self.logger.info("Received message from host")
+        self.logger.debug("Received:\n%s", raw)
+        return super(TLSSession, self)._dispatch_message(raw)
+
+    def close(self):
+        self._closing.set()
+        self._socket.close()
+        self._connected = False
+
+    def connect(self, host=None, port=DEFAULT_TLS_NETCONF_PORT,
+                keyfile=None, certfile=None, ca_certs=None,
+                timeout=DEFAULT_TLS_TIMEOUT):
+        """Establish NETCONF session via TLS.
+
+        :param host: Hostname or IP address to connect to.
+        :param port: Port number which would be used for connection, the
+            default value is 6513 (`DEFAULT_TLS_NETCONF_PORT`).
+        :param certfile: Points to file in PEM format with client's certificate
+            and optionally with client's private key.
+        :param keyfile: Points to file with client's private key in PEM format,
+            if the latter is not in certfile.
+        :param ca_certs: Points to file containing a list of certificates,
+            which should be used to verify server certificate.
+        :param timeout: Specifies the connection timeout, defaults to
+            120 seconds (`DEFAULT_TLS_TIMEOUT`).
+
+        :raise TLSError if the connection can not be established.
+        """
+        if host is None:
+            raise TLSError('Missing host')
+        if certfile is None:
+            raise TLSError('Missing client certificate file')
+
+        ssl_context = SSLContext(PROTOCOL_TLSv1_2)
+        try:
+            ssl_context.load_cert_chain(certfile=certfile, keyfile=keyfile)
+        except SSLError:
+            raise TLSError('Bad client private key / certificate pair')
+        except IOError:
+            raise TLSError('Private key / certificate pair not found')
+
+        if ca_certs:
+            ssl_context.verify_mode = CERT_REQUIRED
+            try:
+                ssl_context.load_verify_locations(cafile=ca_certs)
+            except SSLError:
+                raise TLSError('Bad Certification Authority file')
+            except IOError:
+                raise TLSError('CA certificate file not found')
+
+        sock = socket.socket(AF_INET, SOCK_STREAM)
+        ssl_sock = ssl_context.wrap_socket(sock, do_handshake_on_connect=False)
```

Comment:
When user chooses to use `PROTOCOL_TLS_CLIENT` protocol, it automatically enables the hostname checking (sets `SSLContext.check_hostname` to  `True`). With checking enabled, the `wrap_socket` method expects `server_hostname` argument, which is usually same as the `host`, but might be useful to have is an separate argument which defaults to `host` value when not provided. 

`def connect(self, ..., server_hostname=None):`
 
 and then
 
 `ssl_sock = ssl_context.wrap_socket(sock, do_handshake_on_connect=False, server_hostname=(server_hostname or host)`
 
 [Here's link to the doc which mentions this behavior.](https://docs.python.org/3/library/ssl.html#ssl.SSLContext.check_hostname)

Reply:
Ok, added it! This also gave me an idea to introduce `check_hostname` parameter to disable (or enable) hostname checking in the first place, because before it was explicitly disabled.